### DISCUSSION TITLE
fcitx5 support for KDE Plasma 5.21

### DISCUSSION THREAD

**Discussion by Dyrimon** [OpenBangla/OpenBangla-Keyboard#213] [Q&A] [answered]
Hi, I'm running Fedora 34 with Plasma 5.21.4 . Plasma has horrible support for `ibus`, on the point that it's just far less painful to use `fcitx5` . Following the [instructions](https://github.com/OpenBangla/OpenBangla-Keyboard/wiki/Configure-on-KDE) for KDE doesn't work. There is no "ibus-preferences" to work with and running `ibus-setup` from terminal will work only for that session. I've added the environment configs in `.bashrc` and added an autostart config for `ibus-daemon` . Also, ibus frequently crashes on boot and refuses to run even if started manually. If, fortunately ibus does run, you have to switch input method with mouse as keyboard shortcuts won't work. After that it will give you a popup that is annoying to get rid of. I hope fcitx5 support will soon be available.

**Comment by mominul** [comment, marked as answer]
I have just merged the `fcitx` integration into the [`develop`](https://github.com/OpenBangla/OpenBangla-Keyboard/tree/develop) branch. Do you want to check that out? We'll love early bug reports! :grin:

**Reply by Dyrimon** [reply]
Hi, I already have `openbangla-keyboard` installed. How do I add the develop branch? Do I have to uninstall it completely and clone the git repo?

**Reply by mominul** [reply]
Yeah, you have to uninstall `openbangla-keyboard` & `ibus` first.
After that, you have to run these following commands:
```
sudo dnf install @buildsys-build rust cargo cmake qt5-qtdeclarative-devel libzstd-devel fcitx5 fcitx5-devel
git clone --recursive https://github.com/OpenBangla/OpenBangla-Keyboard.git
cd OpenBangla-Keyboard
git checkout develop
git submodule update
mkdir build && cd build
cmake ..
make
sudo make install
```
After that, log out from your session. Next steps are [avalilable here](https://gist.github.com/mominul/2dd0de850ad6a74271a1475a2b93ef78#%E0%A6%93%E0%A6%AA%E0%A7%87%E0%A6%A8%E0%A6%AC%E0%A6%BE%E0%A6%82%E0%A6%B2%E0%A6%BE-%E0%A6%95%E0%A6%BF%E0%A6%AC%E0%A7%8B%E0%A6%B0%E0%A7%8D%E0%A6%A1-%E0%A6%AF%E0%A7%81%E0%A6%95%E0%A7%8D%E0%A6%A4-%E0%A6%95%E0%A6%B0%E0%A6%BE).

**Reply by Dyrimon** [reply]
OK, so far I've failed to build it. I don't have numpad on my laptop so I'm guessing there are some hard keybinds that are throwing this error.

Here is my build log: https://bin.privacytools.io/?19698cec96855245#E2hijRKx17Z1vxYnYjWGTvN5RzpizaeM1pycmtdYHNYw

Where did I screw up?

**Reply by mominul** [reply]
Actually, I was the one who screwed up :sweat_smile: 
One has to run `git submodule update` after `git checkout develop` because the branches point to different commits of the submodule, I have edited my comment to reflect this.
Fear not, you don't have to go through the whole process again! :grin:  Just `cd` to the `OpenBangla-Keyboard/build` directory and run these following commands:
```
git submodule update
cmake ..
make
sudo make install
```

Sorry for the trouble!

**Reply by Dyrimon** [reply]
So far so good! Working as expected ~~but the layout switching shortcut seems stuck to Ctrl+Space. It does not respect the default Plasma settings (`Ctrl+Alt+K` or `Meta+Space` whichever it is)~~. Also the font is very thin. Is it possible to install extra Bangla Font and make OBK choose it? Thanks.

**Reply by Dyrimon** [reply]
My bad, I had `fcitx5` hotkey and `kwin` shortcut mixed up! 😞. Plasma settings are confusing as hell and I'm still getting used to it.

**Comment by mominul** [comment]
> Also the font is very thin. Is it possible to install extra Bangla Font and make OBK choose it?

I have also noticed it. I am not a KDE or fcitx user, so I can't help in this matter. Fcitx is [quite customizable](https://wiki.archlinux.org/title/Fcitx5#Configuration), so there must be a way to change the font. cc @krafi99

**Comment by Dyrimon** [comment]
It seems OpenBangla still requires `ibus` even if it uses `fcitx`. Building on openSUSE Tumblewedd shows error demanding `ibus`
```
cmake ..
-- The C compiler identification is GNU 11.1.1
-- The CXX compiler identification is GNU 11.1.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Cargo Home: /home/geeko/.cargo
-- Rust Compiler Version: rustc 1.52.1
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.7.3") 
-- Checking for module 'ibus-1.0'
--   Package 'ibus-1.0', required by 'virtual:world', not found
CMake Error at /usr/share/cmake/Modules/FindPkgConfig.cmake:556 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPkgConfig.cmake:778 (_pkg_check_modules_internal)
  CMakeLists.txt:78 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
```

**Reply by ahmubashshir** [reply]
disable ibus support while building.
https://github.com/OpenBangla/OpenBangla-Keyboard/pull/222#issue-645742537

**Reply by mominul** [reply]
@Dyrimon  You have to disable `ibus` and enable `fcitx` support like this:
```
cmake .. -DENABLE_IBUS=OFF -DENABLE_FCITX=ON
```

**Reply by Dyrimon** [reply]
cmake is still having problems, now searching for `Fcitx5Core` , which is not in the openSUSE repo
```
CMake Error at CMakeLists.txt:83 (find_package):
  By not providing "FindFcitx5Core.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "Fcitx5Core", but CMake did not find one.

  Could not find a package configuration file provided by "Fcitx5Core"
  (requested version 5.0.5) with any of the following names:

    Fcitx5CoreConfig.cmake
    fcitx5core-config.cmake

  Add the installation prefix of "Fcitx5Core" to CMAKE_PREFIX_PATH or set
  "Fcitx5Core_DIR" to a directory containing one of the above files.  If
  "Fcitx5Core" provides a separate development package or SDK, be sure it has
  been installed.

-- Configuring incomplete, errors occurred!
```
I took a look at CMakeOutput.log but it gives me no hint at the problem
[CMakeOutput.log](https://github.com/OpenBangla/OpenBangla-Keyboard/files/6690975/CMakeOutput.log)

**Reply by Dyrimon** [reply]
I think I'll just install ibus and it's dependencies to get it over with.

**Reply by mominul** [reply]
This error is because cmake is unable to find the `fcitx` development package. Have you installed it?

**Reply by ahmubashshir** [reply]
@Dyrimon install [fcitx5-modules-dev](https://pkgs.org/download/fcitx5-modules-dev)

**Reply by Dyrimon** [reply]
> @Dyrimon install [fcitx5-modules-dev](https://pkgs.org/download/fcitx5-modules-dev)

The link only gives binaries for Debian/Ubuntu. This package does not exist in openSUSE repos, actually problem is the packages are named as `fcitx` including related packages rather than `fcitx5`. I think this OBK rpm packaging is built in favor of Fedora, where such naming convention exists. I've already installed all the dependencies related to `fcitx`, including qt5 related packages.

**Reply by mominul** [reply]
@Dyrimon I haven't used openSUSE before but I think you will need to install these two packages [`libFcitx5Core7`](https://software.opensuse.org/package/libFcitx5Core7) and [`libFcitx5Utils2`](https://software.opensuse.org/package/libFcitx5Utils2).

**Reply by Dyrimon** [reply]
> @Dyrimon I haven't used openSUSE before but I think you will need to install these two packages [`libFcitx5Core7`](https://software.opensuse.org/package/libFcitx5Core7) and [`libFcitx5Utils2`](https://software.opensuse.org/package/libFcitx5Utils2).

Installing these packages from the untested third-party repos breaks and conflicts with installed `fcitx` packages from the official repo. I'm going to give up on this for now and wait for the fcitx support properly merged. Installing the default ibus and related dependencies also doesn't work.

**Reply by mominul** [reply]
Have you installed the [`fcitx5`, version 5.0.8](https://software.opensuse.org/package/fcitx5) package? `fcitx` package includes the old 4.* version of Fcitx. OBK needs version 5 of Fcitx.
> wait for the fcitx support properly merged

Fcitx is support is properly merged and in working condition, you are just having issues because of the bad state of `fcitx5` packages in openSUSE.

**Reply by Dyrimon** [reply]
> Have you installed the [`fcitx5`, version 5.0.8](https://software.opensuse.org/package/fcitx5) package? `fcitx` package includes the old 4.* version of Fcitx. OBK needs version 5 of Fcitx.

I caved in and finally installed the third-party packages. For some reason I had to `make` and install twice to get it worked 😆 .
I also made a migration from fcitx4 to fcitx5 using the wizard for good measure. Now it's working 🎉 . I guess I'm just annoyed that I had to break existing packages and install untested ones. But anyway I'll end my rant. Thank you both @mominul and @ahmubashshir for all the help! ❤️

**Reply by mominul** [reply]
Glad to help! 😊

**Comment by ahmubashshir** [comment]
You're welcome :3
	Mubashshir

**Comment by rashidi77** [comment]
FCITX 5 is now up on Flathub. Is it somehow possible to integrate OpenBangla Keyboard with this Flatpak? It will be very useful for SteamDeck and EndlessOS users. 

Flatpak menifest for FCITX 5 : 
https://github.com/flathub/org.fcitx.Fcitx5

Flatpak download link : 
https://beta.flathub.org/apps/details/org.fcitx.Fcitx5

Their WIKI also mentions the Flatpak version : 
https://fcitx-im.org/wiki/Install_Fcitx_5#Install_Fcitx_5_from_Flatpak

**Comment by ysahmed** [comment]
I have compiled and installed `OBK fcixt5` but when I switch to `OBK` it says `OpenBangla keyboard (not available)`
- fedora 40
- KDE 6.1
